Add support for multiple Twitter Accounts

iamdharmesh opened the pull request:
### Description of the Change
Added support for connecting multiple Twitter Accounts and sending tweets to those accounts.

PR also, Fixes E2E tests failure by skip Twitter API calls.

**Block Editor:**
| New Post | Published Post |
| --- | --- |
|<img width="1039" alt="image" src="https://github.com/10up/autoshare-for-twitter/assets/10613171/cbfe9b85-d694-46ab-b4e6-15b30cf5754c">|<img width="1039" alt="image" src="https://github.com/10up/autoshare-for-twitter/assets/10613171/cbb70a7a-c99f-494f-aae3-6993cfe3c8bd">|

**Classic Editor:**
| New Post | Published Post |
| --- | --- |
|<img width="1039" alt="image" src="https://github.com/10up/autoshare-for-twitter/assets/10613171/5924597b-7817-4657-b4ce-8a1f055208ff">|<img width="1039" alt="image" src="https://github.com/10up/autoshare-for-twitter/assets/10613171/80d67783-1c1e-4549-a40d-1c151f4e9bcc">|

|**Settings:**|
|---|
|<img width="1173" alt="image" src="https://github.com/10up/autoshare-for-twitter/assets/10613171/a7d8cbc1-fe5d-4681-b3bc-41077c7a7373">|


Closes #230 
Closes #237 

### How to test the Change
**Backward compatibility**
1. Install and Configure current version of plugin.
2. Checkout this PR
3. Visit "Autoshare for Twitter" settings page or "Add new" post page.
4. You should see 1 account connected and UI for enable tweet on it.

**New features**
1. Verify that you are able to configure twitter credentials by following setup instructions
1. After Twitter app credentials saved, UI for connect Twitter account is visible.
1. Verify that you are able to connect more than 1 accounts.
1. Verify that you are able to disconnect accounts
1. Publish new post and make sure it send tweet to only tweet enabled accounts.
1. Configure "Autoshare by default" settings for each account and verify that from create new post screen
1. Follow steps 3-6 for classic editor and block editor both.


### Changelog Entry
> Added - Support for connect multiple Twitter accounts
> Fixed - E2E tests by skip calling Twitter API calls


### Credits
<!-- Please list any and all contributors on this PR so that they can be added to this projects CREDITS.md file. -->
Props @iamdharmesh @jeffpaul 


### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you are unsure about any of these, please ask for clarification.  We are here to help! -->
- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests pass.

iamdharmesh committed:
Added Twitter accounts list table.

iamdharmesh committed:
Added Twitter Auth class.

iamdharmesh committed:
Added utils functions.

iamdharmesh committed:
Added accounts list table in settings screen.

iamdharmesh committed:
Added Twitter API class to handle API requests.

iamdharmesh committed:
Update settings section to add twitter accounts UI

iamdharmesh committed:
Added account toggle UI in gutenberg.

iamdharmesh committed:
Merge branch 'develop' of github.com:10up/autoshare-for-twitter into enhancement/230

iamdharmesh committed:
Fix PHP Unit tests.

iamdharmesh committed:
Handled tweet to multiple account.

iamdharmesh committed:
Created a component for status logs.

iamdharmesh committed:
Update status log on classic editor.

iamdharmesh committed:
Added twitter accounts UI in classic editor.

iamdharmesh committed:
Fix styles.

iamdharmesh committed:
Added disconnect account functionality.

iamdharmesh committed:
Don't load panel if we don't have any accounts connected.

iamdharmesh committed:
Fix incorrect conditions

iamdharmesh committed:
Added aushare for default setting for each connected account.

iamdharmesh committed:
Enabled default autoshare in classic editor.

iamdharmesh committed:
Added test plugin to skip twitter API calls.

iamdharmesh committed:
Updated admin tests as per new settings.

iamdharmesh committed:
Added multi accounts e2e tests.

iamdharmesh committed:
Update existing e2e editor tests.

iamdharmesh committed:
Fix unit tests.

iamdharmesh committed:
Update E2E worflow file.

iamdharmesh committed:
Fix plugin path in e2e workflow file.

iamdharmesh committed:
Mock disconnect request.

iamdharmesh committed:
Added tests for account check/uncheck.

iamdharmesh committed:
Added Disconnect tests and added few commands to increase code reusability.

iamdharmesh committed:
Fix flaky tests

iamdharmesh committed:
Fixed flaky e2e tests.

iamdharmesh committed:
Fixed unit tests.

iamdharmesh committed:
Revert "Fixed flaky e2e tests."

This reverts commit 6b331dd6625a8165bb34c5d6ec5b39c6fb8e81ac.

iamdharmesh committed:
Different approach.

iamdharmesh committed:
Added retry mechanism in run mode.

iamdharmesh committed:
Update E2E tests.

iamdharmesh committed:
EOF lines added and indentation changes.

Sidsector9 commented on includes/admin/post-meta.php:
```diff
@@ -262,6 +277,55 @@ function render_tweet_submitbox( $post ) {
+			<div class="twitter-account-wrapper">
+				<img src="<?php echo esc_url( $account['profile_image_url'] ); ?>" alt="<?php echo esc_attr( $account['name'] ); ?>" class="twitter-account-profile-image" />
+				<span class="account-details">
+					<strong>@<?php echo esc_attr( $account['username'] ); ?></strong>
```
Should be `esc_html()`

Sidsector9 commented on includes/admin/post-meta.php:
```diff
@@ -262,6 +277,55 @@ function render_tweet_submitbox( $post ) {
+				<span class="account-details">
+					<strong>@<?php echo esc_attr( $account['username'] ); ?></strong>
+					<br />
+					<?php echo esc_attr( $account['name'] ); ?>
```
Should be `esc_html()`

Sidsector9 commented on includes/admin/post-meta.php:
```diff
@@ -262,6 +277,55 @@ function render_tweet_submitbox( $post ) {
+		?>
+		<span class="connect-account-link">
+			<a href="<?php echo esc_url( admin_url( 'options-general.php?page=autoshare-for-twitter' ) ); ?>" target="_blank">
+				<?php esc_attr_e( 'Connect an account', 'autoshare-for-twitter' ); ?>
```
Should be `esc_html_e()`

Sidsector9 commented on includes/admin/post-meta.php:
```diff
@@ -414,14 +484,16 @@ function get_tweet_status_logs( $post ) {
 		esc_url( $twitter_url ),
-		esc_html__( 'View', 'autoshare-for-twitter' )
+		esc_html__( 'View', 'autoshare-for-twitter' ),
+		esc_attr( $handle )
```
I've observed this pattern where the handle is escaped using esc_attr() instead of esc_html(). Is it because the handle accepts limited characters?

iamdharmesh replied:
Not fully aware on limited characters, I just updated all escape function to `esc_html`. Thanks for adding note for this.

Sidsector9 commented on includes/class-twitter-accounts.php:
```diff
@@ -0,0 +1,331 @@
+
+		// Check if the nonce is valid.
+		if ( ! isset( $_GET['autoshare_twitter_authorize_nonce'] ) || ! wp_verify_nonce( sanitize_text_field( wp_unslash( $_GET['autoshare_twitter_authorize_nonce'] ) ), 'autoshare_twitter_authorize_action' ) ) {
+			wp_die( esc_html__( 'You have not access to doing this operations.', 'autoshare-for-twitter' ) );
```
This could be:

"You are not authorised to perform this operation"

iamdharmesh committed:
Merge branch 'develop' of github.com:10up/autoshare-for-twitter into enhancement/230

Sidsector9 commented on includes/admin/post-transition.php:
```diff
@@ -42,7 +43,7 @@ function setup() {
  * @param WP_Post $post       The current post.
  *
- * @return object
+ * @return null
```
Should be `void`

Sidsector9 commented on includes/admin/post-transition.php:
```diff
@@ -92,7 +94,7 @@ function( $post ) {
  * @param bool $force   Publish tweet regardless of autoshare enabled or disabled on post.
  *
- * @return object
+ * @return bool|null
```
Can you confirm if this function returns `null`?

iamdharmesh replied:
updated to `bool|void`

Sidsector9 commented on includes/class-twitter-accounts.php:
```diff
@@ -0,0 +1,331 @@
+			// phpcs:ignore WordPress.Security.NonceVerification.Recommended
+			$is_deniend = isset( $_GET['denied'] ) ? sanitize_text_field( wp_unslash( $_GET['denied'] ) ) : false;
+			if ( $is_deniend ) {
+				throw new \Exception( __( 'You have denied the authorization.', 'autoshare-for-twitter' ) );
```
@iamdharmesh Correct me if I'm wrong, by the code it seems that the API returns with this error.
This can be "Authorization denied for this request."

Sidsector9 commented on includes/class-twitter-accounts.php:
```diff
@@ -0,0 +1,331 @@
+			$access_token   = $this->twitter_api->get_access_token( $oauth_token, $oauth_token_secret, $oauth_verifier );
+
+			if ( ! $access_token || ! isset( $access_token['oauth_token'] ) || ! isset( $access_token['oauth_token_secret'] ) ) {
+				throw new \Exception( 'autoshare_twitter_authorize_error', __( 'Something went wrong during getting access token. Please try again', 'autoshare-for-twitter' ) );
```
This can be "Something went wrong while generating the access token" or "Something went wrong while fetching the access token"

Sidsector9 commented on includes/class-twitter-accounts.php:
```diff
@@ -0,0 +1,331 @@
+
+		// Check if the nonce is valid.
+		if ( ! isset( $_GET['autoshare_twitter_disconnect_nonce'] ) || ! wp_verify_nonce( sanitize_text_field( wp_unslash( $_GET['autoshare_twitter_disconnect_nonce'] ) ), 'autoshare_twitter_disconnect_action' ) ) {
+			wp_die( esc_html__( 'You have not access to doing this operations.', 'autoshare-for-twitter' ) );
```
"You are not authorised to perform this operation"

Sidsector9 commented on includes/class-twitter-accounts.php:
```diff
@@ -0,0 +1,331 @@
+		}
+
+		if ( ! isset( $_GET['account_id'] ) ) {
+			wp_die( esc_html__( 'Twitter account ID is required to doing this operation.', 'autoshare-for-twitter' ) );
```
Twitter account ID is required to perform this operation.

Sidsector9 reviewed (changes requested):
@iamdharmesh thanks for this amazing work! I've requested for few changes, let me know if you'd like to discuss it over a call.

iamdharmesh committed:
PR Feedback.

iamdharmesh commented:
Thanks a lot for the review and feedback @Sidsector9. I have addressed all feedback in [5a24c7d](https://github.com/10up/autoshare-for-twitter/pull/238/commits/5a24c7d882107d45129ca792739424bd79a529dc). Could you please help to re-review and confirm once?

Thanks.

Sidsector9 commented on includes/admin/post-transition.php:
```diff
@@ -92,7 +94,7 @@ function( $post ) {
  * @param bool $force   Publish tweet regardless of autoshare enabled or disabled on post.
  *
- * @return object
+ * @return bool|void
```
@iamdharmesh I think we should strictly set it to `bool`.

In case this function returns `void` which will evaluate to `null`, that `null` will be set here:

https://github.com/10up/autoshare-for-twitter/blob/73ab6c80061ac3862a16909a9c98ebfee71d836a/includes/admin/post-transition.php#L171

In JS, we strictly check it against `false` here:

https://github.com/10up/autoshare-for-twitter/blob/73ab6c80061ac3862a16909a9c98ebfee71d836a/assets/js/admin-autoshare-for-twitter-classic-editor.js#L218

What do you think?

iamdharmesh replied:
Yeah, not related to this PR but important enough to fix it here. Updated. Thanks

iamdharmesh committed:
Update return type to bool on publish_tweet function.
